File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.py
from collections import deque

def solution(maps):
    answer = 0
    N, M = len(maps), len(maps[0])
    start_r, start_c, lever_r, lever_c, end_r, end_c = 0, 0, 0, 0, 0, 0
    for i in range(N):
        for j in range(M):
            if maps[i][j] == "S":
                start_r, start_c = i, j
            elif maps[i][j] == "L":
                lever_r, lever_c = i, j
            elif maps[i][j] == "E":
                end_r, end_c = i, j
    
    start_to_lever = bfs(start_r, start_c, lever_r, lever_c, maps)
    # print(start_to_lever)
    if start_to_lever == -1:
        return -1
    
    lever_to_end = bfs(lever_r, lever_c, end_r, end_c, maps)
    # print(lever_to_end)
    if lever_to_end == -1:
        return -1
    
    return start_to_lever + lever_to_end

def bfs(start_r, start_c, end_r, end_c, maps):
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    
    N, M = len(maps), len(maps[0])
    visited = [[False for _ in range(M)] for __ in range(N)]
    dq = deque()
    dq.append([start_r, start_c, 0])
    visited[start_r][start_c] = True
    
    while dq:
        cr, cc, cd = dq.popleft()
        # print(cr, cc, cd)
        if cr == end_r and cc == end_c:
            return cd
        
        for i in range(4):
            nr, nc = cr + dr[i], cc + dc[i]
            if not (0 <= nr < N) or not (0 <= nc < M) or visited[nr][nc] or maps[nr][nc] == "X":
                continue
            dq.append([nr, nc, cd + 1])
            visited[nr][nc] = True
        
    return -1
    

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    int[] dr = {-1, 1, 0, 0};
    int[] dc = {0, 0, -1, 1};
    public int solution(String[] maps) {
        int startR = 0, startC = 0;
        int leverR = 0, leverC = 0;
        int exitR = 0, exitC = 0;
            
        for(int r = 0; r < maps.length; r++){
            for(int c = 0; c < maps[r].length(); c++){
                if(maps[r].charAt(c) == 'S'){
                    startR = r;
                    startC = c;
                } else if(maps[r].charAt(c) == 'L'){
                    leverR = r;
                    leverC = c;
                } else if(maps[r].charAt(c) == 'E'){
                    exitR = r;
                    exitC = c;
                }
            }
        }                        
        
        int startToLever = bfs(startR, startC, leverR, leverC, maps);
        if(startToLever == -1) return -1;
        
        int leverToExit = bfs(leverR, leverC, exitR, exitC, maps);
        if(leverToExit == -1) return -1;
        
        return startToLever + leverToExit;
    }
    
    public int bfs(int startR, int startC, int endR, int endC, String[] maps){
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[] {startR, startC, 0});
        
        int R = maps.length;
        int C = maps[0].length();
        
        boolean[][] visited = new boolean[R][C];
        visited[startR][startC] = true;
        
        while(!queue.isEmpty()){
            int curR = queue.peek()[0];
            int curC = queue.peek()[1];
            int curD = queue.peek()[2];
            queue.poll();
            
            if(curR == endR && curC == endC){
                return curD;
            }
            
            for(int d = 0; d < 4; d++){
                int nextR = curR + dr[d];
                int nextC = curC + dc[d];
                
                if(nextR < 0 || R <= nextR || nextC < 0 || C <= nextC || maps[nextR].charAt(nextC) == 'X'|| visited[nextR][nextC]) continue;
                queue.add(new int[] {nextR, nextC, curD + 1});
                visited[nextR][nextC] = true;
            }
        }
        
        return -1;
    }
}